File: 02_LearnDSP-MATLAB/Sec02_FrequencyAnalysis/case02_FrequencySpectrum.m
% Plot frequency spectrum to analysis singal in frequency domain
% Key: align to frequency points

% XiaoCY 2021-02-02

%% time domain
clear;clc

fs = 1000;                      % sampling frequency
fsig = 100;                     % signal frequency
fext = 250;                     % extra frequency (noise)

t = (0:1/fs:100/fsig)';         % time series
N = length(t);                  % number of samples

x = cos(2*pi*fsig*t)+0.2*sin(2*pi*fext*t);

figure('Name','time')
plot(t,x)
grid on
xlabel('Time (s)')
ylabel('Signal')

%% frequency domain
X = fft(x);

Nf = floor(N/2);                % number of frequency points
f = (0:Nf)'*fs/N;

amp = abs(X(1:Nf+1))*2/N;
phase = angle(X(1:Nf+1));


figure('Name','FFT')            % show real part and imaginary part
subplot(2,1,1)
plot(f,real(X(1:Nf+1)))
grid on
xlabel('Frequency (Hz)')
ylabel('Re(X)')

subplot(2,1,2)
plot(f,imag(X(1:Nf+1)))
grid on
xlabel('Frequency (Hz)')
ylabel('Im(X)')


figure('Name','spectrum')       % show frequency spectrum
subplot(2,1,1)
plot(f,amp)
grid on
xlabel('Frequency (Hz)')
ylabel('Amplitude')

subplot(2,1,2)
plot(f,phase)
grid on
xlabel('Frequency (Hz)')
ylabel('Phase (rad)')

File: 03_LearnDSP-Python/Sec02_FrequencyAnalysis/case02_FrequencySpectrum.py
'''
Plot frequency spectrum to analysis signal in frequecy domain
Key: align to frequency points

XiaoCY 2021-02-02
'''

#%%
import numpy as np
import matplotlib.pyplot as plt

#%% time domain
fs = 1000                       # sampling frequency
fsig = 100                      # signal frequency
fext = 250                      # extra frequency (noise)

t = np.arange(0,100/fsig,1/fs)  # time series
N = len(t)                      # number of samples

x = np.cos(2*np.pi*fsig*t)+0.2*np.sin(2*np.pi*fext*t)

plt.figure(1)
plt.plot(t,x)
plt.grid()
plt.title('Time Domain')
plt.xlabel('Time (s)')
plt.ylabel('Signal')

#%% frequency domain
X = np.fft.fft(x,axis=0)

Nf = int(np.ceil(N/2))
f = np.arange(0,Nf)*fs/N       # frequency points

amp = np.abs(X[0:Nf])*2/N
phase = np.angle(X[0:Nf])

plt.figure(2)
plt.subplot(2,1,1)
plt.plot(f,X[0:Nf].real)
plt.grid()
plt.xlabel('Frequency (Hz)')
plt.ylabel('Re(X)')

plt.subplot(2,1,2)
plt.plot(f,X[0:Nf].imag)
plt.grid()
plt.xlabel('Frequency (Hz)')
plt.ylabel('Im(X)')
plt.show

plt.figure(3)
plt.subplot(2,1,1)
plt.plot(f,amp)
plt.grid()
plt.xlabel('Frequency (Hz)')
plt.ylabel('Amplitude')

plt.subplot(2,1,2)
plt.plot(f,phase)
plt.grid()
plt.xlabel('Frequency (Hz)')
plt.ylabel('Phase (rad)')